### ISSUE TITLE
imagenet resnet pytorch workload CUBLAS_STATUS_NOT_INITIALIZED 

### ISSUE THREAD

**Issue by priyakasimbeg** [mlcommons/algorithmic-efficiency#185]
Running the imagnet resnet pytorch workload results in an error that may indicate an oom issue.

## Description

Traceback:

```
I1013 18:57:50.048486 139836274620224 submission_runner.py:231] Initializing model.
I1013 18:57:52.236851 139836274620224 submission_runner.py:234] Initializing optimizer.
I1013 18:57:52.331568 139836274620224 submission_runner.py:243] Initializing metrics bundle.
I1013 18:57:52.331735 139836274620224 submission_runner.py:257] Starting training loop.
Traceback (most recent call last):
  File "submission_runner.py", line 495, in <module>
    app.run(main)
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/absl/app.py", line 312, in run
    _run_main(main, args)
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/absl/app.py", line 258, in _run_main
    sys.exit(main(argv))
  File "submission_runner.py", line 479, in main
    FLAGS.tokenizer_vocab_path)
  File "submission_runner.py", line 411, in score_submission_on_workload
    tokenizer_vocab_path)
  File "submission_runner.py", line 285, in train_once
    rng=update_rng)
  File "/home/kasimbeg/algorithmic-efficiency/target_setting_runs/pytorch_submission_base.py", line 38, in update_params
    update_batch_norm=True)
  File "/home/kasimbeg/algorithmic-efficiency/algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/workload.py", line 191, in model_fn
    logits_batch = model(augmented_and_preprocessed_input_batch['inputs'])
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 1008, in forward
    output = self._run_ddp_forward(*inputs, **kwargs)
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 969, in _run_ddp_forward
    return module_to_run(*inputs[0], **kwargs[0])
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/kasimbeg/algorithmic-efficiency/algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/models.py", line 266, in forward
    x = self.fc(x)
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`
```

## Steps to Reproduce
On vm kasimbeg-5 run:
```
torchrun --redirects 1:0,2:0,2:0,4:0,5:0,6:0,7:0 --standalone --nnodes=1 --nproc_per_node=8 submission_runner.py     --framework=pytorch     --data_dir=~/local_data/imagenet/pytorch     --imagenet_v2_data_dir=~     --experiment_dir=~     --experiment_name=target_setting     --workload=imagenet_resnet     --submission_path=target_setting_runs/pytorch_nesterov.py     --tuning_search_space=target_setting_runs/imagenet_resnet/tuning_search_space.json
```
Note you'll probably have to modify the `--tuning_search_space` path since the files have been reorganized recent [pull/174](https://github.com/mlcommons/algorithmic-efficiency/pull/174)

## Source or Possible Fix
Could be an OOM issue, see [post](https://discuss.pytorch.org/t/cuda-error-cublas-status-not-initialized-when-calling-cublascreate-handle/125450/2)

**Comment by priyakasimbeg** [comment]
@runame did you have the script with the environment set up ready by any chance? 
It may fix this issue as well (hopefully).

**Comment by runame** [comment]
Have you already tried it with PR #190 or the main branch after it was merged?

**Comment by priyakasimbeg** [comment]
I just ran this on the up to date main branch and I still get this error:

```
I1021 20:37:19.412386 139692541118272 submission_runner.py:239] Initializing checkpoint and logger.                                                                                
I1021 20:37:20.155205 139692541118272 submission_runner.py:260] Saving meta data to /home/kasimbeg/target_setting/imagenet_resnet_pytorch/trial_1/meta_data_0.json                 
I1021 20:37:20.156133 139692541118272 submission_runner.py:263] Saving flags to /home/kasimbeg/target_setting/imagenet_resnet_pytorch/trial_1/flags_0.json                         
I1021 20:37:20.290321 139692541118272 submission_runner.py:269] Starting training loop.                                                                                            
Traceback (most recent call last):                                                                                                                                                 
  File "submission_runner.py", line 547, in <module>                                                                                                                               
    app.run(main)                                                                                                                                                                  
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/absl/app.py", line 312, in run                                                                       
    _run_main(main, args)                                                                                                                                                          
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/absl/app.py", line 258, in _run_main                                                                 
    sys.exit(main(argv))                                                                                                                                                           
  File "submission_runner.py", line 531, in main                                                                                                                                   
    FLAGS.tokenizer_vocab_path)                                                                                                                                                    
  File "submission_runner.py", line 466, in score_submission_on_workload                                                                                                           
    tokenizer_vocab_path)                                                                                                                                                          
  File "submission_runner.py", line 299, in train_once                                                                                                                             
    rng=update_rng)                                                                                                                                                                
  File "/home/kasimbeg/algorithmic-efficiency/reference_algorithms/target_setting_algorithms/pytorch_submission_base.py", line 40, in update_params                                
    update_batch_norm=True)                                                                                                                                                        
  File "/home/kasimbeg/algorithmic-efficiency/algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/workload.py", line 196, in model_fn                                
    logits_batch = model(augmented_and_preprocessed_input_batch['inputs'])                                                                                                         
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                                                
    return forward_call(*input, **kwargs)                                                                                                                                          
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 1008, in forward                                             
    output = self._run_ddp_forward(*inputs, **kwargs)                                                                                                                              
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 969, in _run_ddp_forward                                     
    return module_to_run(*inputs[0], **kwargs[0])                                                                                                                                  
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                                                
    return forward_call(*input, **kwargs)                                                                                                                                          
  File "/home/kasimbeg/algorithmic-efficiency/algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/models.py", line 270, in forward                                   
    x = self.fc(x)                                                                                                                                                                 
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                                                
    return forward_call(*input, **kwargs)                                                                                                                                          
  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 114, in forward                                                    
    return F.linear(input, self.weight, self.bias)                                                                                                                                 
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)` 
```

**Comment by priyakasimbeg** [comment]
Not sure what is going on. I got a complete run that was started on Wed morning.

**Comment by pomonam** [comment]
I also get this error and this error resolves when I set the batch size to 512. Interestingly, on a single GPU run, it does not cause OOM issues too. I will look into this as well.

**Comment by pomonam** [comment]
I looked into this issue. It seems like both increasing or decreasing batch size (e.g., 1024->800, 1024->1312) seem to solve the issue and GPU utilization is maxed at 80%. This means that it is not the OOM issue (OOM happens if we make batch size much larger i.e., 2048). 

I find it very strange that some specific batch size can be the core reason for this error. I tried CUDA 11.6 and this error disappears, so I suspect it is some CUDA issue.
